drivers/at86rf215/at86rf215.c

```diff
@@ -0,0 +1,340 @@
+    at86rf215_reg_write(dev, RG_RF_XOC, AT86RF215_TRIM_VAL | XOC_FS_MASK);
+#endif
+
+    /* enable TXFE & RXFE IRQ */
+    at86rf215_reg_write(dev, dev->BBC->RG_IRQM, BB_IRQ_TXFE | BB_IRQ_RXFE);
+
+    /* enable EDC IRQ */
+    at86rf215_reg_write(dev, dev->RF->RG_IRQM, RF_IRQ_EDC | RF_IRQ_TRXRDY);
+
+    /* set energy detect thresholt to -84 dBm */
+    at86rf215_reg_write(dev, dev->BBC->RG_AMEDT, -84);
+
+    /* enable address filter 0 */
+    at86rf215_reg_write(dev, dev->BBC->RG_AFC0, 0x1);
+    at86rf215_reg_write(dev, dev->BBC->RG_AMAACKPD, 0x1);
+
+    /* enable auto-ACK with Frame Checksum & Data Rate derived from RX frame */
+    at86rf215_reg_write(dev, dev->BBC->RG_AMCS, AMCS_AACK_MASK
+                                              | AMCS_AACKFA_MASK
+                                              | AMCS_AACKDR_MASK);
+
+    at86rf215_configure_legacy_OQPSK(dev, 0);
+
+    /* set default channel */
+    at86rf215_set_chan(dev, dev->netdev.chan);
+
+    /* set short and long address */
+    uint64_t long_addr;
+    memcpy(&long_addr, dev->netdev.long_addr, sizeof(long_addr));
+    at86rf215_set_addr_long(dev, long_addr);
+    at86rf215_set_addr_short(dev, 0, unaligned_get_u16(dev->netdev.short_addr));
+
+    /*** set default PAN id ***/
+    at86rf215_set_pan(dev, 0, dev->netdev.pan);
+
+    /* set default TX power */
+    at86rf215_set_txpower(dev, AT86RF215_DEFAULT_TXPOWER);
+
+    /* start listening for incoming packets */
+    at86rf215_rf_cmd(dev, CMD_RF_RX);
+    at86rf215_await_state(dev, RF_STATE_RX);
+
+    dev->state = AT86RF215_STATE_IDLE;
+}
+
+ssize_t at86rf215_send(at86rf215_t *dev, const void *data, size_t len)
+{
+    /* check data length */
+    if (len > AT86RF215_MAX_PKT_LENGTH) {
+        DEBUG("[at86rf215] Error: data to send exceeds max packet size\n");
+        return -EOVERFLOW;
+    }
+
+    if (at86rf215_tx_prepare(dev)) {
+        return -EBUSY;
+    }
+
+    at86rf215_tx_load(dev, data, len, 0);
+    at86rf215_tx_exec(dev);
+    return len;
+}
+
+void at86rf215_tx_done(at86rf215_t *dev)
+{
+    uint8_t amcs = at86rf215_reg_read(dev, dev->BBC->RG_AMCS);
+
+    /* enable AACK, disable TX2RX */
+    amcs &= ~AMCS_TX2RX_MASK;
+    if (dev->flags & AT86RF215_OPT_AUTOACK) {
+        amcs |= AMCS_AACK_MASK;
+    }
+
+    at86rf215_reg_write(dev, dev->BBC->RG_AMCS, amcs);
+}
+
+static bool _tx_ongoing(at86rf215_t *dev)
+{
+    if (dev->flags & AT86RF215_OPT_TX_PENDING) {
+        return true;
+    }
+
+    if (dev->state == AT86RF215_STATE_TX ||
+        dev->state == AT86RF215_STATE_TX_WAIT_ACK) {
+        return true;
+    }
+
+    return false;
+}
+
+/*
+ * As there is no packet queue in RIOT we have to block in send()
+ * when the device is busy sending a previous frame.
+ *
+ * Since both _send() and _isr() are running in the same thread
+ * we have to service radio events while waiting in order to
+ * advance the previous transmission.
+ */
+static void _block_while_busy(at86rf215_t *dev)
```

Comment:
The IRQ would only generate a message to the radio thread.
This function is called by the radio thread to avoid writing to the TX buffer while there is still an ongoing transmission. (#11263 would remove the need for this).

As all the IRQ does is generate a message to the radio thread, the `_isr()` function that would register the state change can not be run - we have to service the interrupt in the busy loop.

So we don't want to handle now invalid interrupt messages when we exit the loop.

Reply:
I'm not very fond of this solution.
But without a queue the only other option to blocking send is dropping packets when the device is busy, and that definitely breaks fragmented packets.